File: resources/js/Utils/profiler.js
/*
 * Profiling dataset: menghitung karakteristik tiap kolom dari baris sungguhan.
 *
 * Padanan `python/core/profiler.py`. Selama engine Python belum tersambung,
 * perhitungannya dilakukan di peramban atas tabel dari `@/data/profiles`.
 * Seluruh halaman analisis membaca hasil profiling ini, jadi angka di grafik,
 * rekomendasi algoritma, dan pilihan target selalu berasal dari sumber yang sama.
 *
 * Kolom identitas (nilai unik untuk setiap baris) ditandai `isIdentifier` dan
 * dikecualikan dari analisis — mengorelasikan nomor transaksi tidak bermakna.
 */

export function mean(values) {
    return values.length
        ? values.reduce((sum, value) => sum + value, 0) / values.length
        : 0;
}

export function stdDev(values) {
    if (values.length < 2) {
        return 0;
    }

    const average = mean(values);
    const variance =
        values.reduce((sum, value) => sum + (value - average) ** 2, 0) /
        (values.length - 1);

    return Math.sqrt(variance);
}

/** Kuantil dengan interpolasi linear pada data yang sudah terurut. */
export function quantile(sorted, fraction) {
    if (!sorted.length) {
        return 0;
    }

    const position = (sorted.length - 1) * fraction;
    const lower = Math.floor(position);
    const upper = Math.ceil(position);

    if (lower === upper) {
        return sorted[lower];
    }

    return sorted[lower] + (sorted[upper] - sorted[lower]) * (position - lower);
}

export function pearson(a, b) {
    const pairs = a
        .map((value, index) => [value, b[index]])
        .filter(([x, y]) => Number.isFinite(x) && Number.isFinite(y));

    if (pairs.length < 3) {
        return 0;
    }

    const xs = pairs.map(([x]) => x);
    const ys = pairs.map(([, y]) => y);
    const meanX = mean(xs);
    const meanY = mean(ys);

    let numerator = 0;
    let sumX = 0;
    let sumY = 0;

    for (let index = 0; index < pairs.length; index += 1) {
        const dx = xs[index] - meanX;
        const dy = ys[index] - meanY;

        numerator += dx * dy;
        sumX += dx * dx;
        sumY += dy * dy;
    }

    const denominator = Math.sqrt(sumX * sumY);

    return denominator === 0 ? 0 : numerator / denominator;
}

export const isNumericType = (type) => type === 'integer' || type === 'float';

/** Angka panjang dipersingkat agar muat sebagai label sumbu. */
export function formatNumber(value) {
    if (!Number.isFinite(value)) {
        return '—';
    }

    const magnitude = Math.abs(value);

    if (magnitude >= 1_000_000_000) {
        return `${(value / 1_000_000_000).toFixed(1).replace('.', ',')} M`;
    }

    if (magnitude >= 1_000_000) {
        return `${(value / 1_000_000).toFixed(1).replace('.', ',')} jt`;
    }

    if (magnitude >= 10_000) {
        return `${Math.round(value / 1000).toLocaleString('id-ID')} rb`;
    }

    if (Number.isInteger(value)) {
        return value.toLocaleString('id-ID');
    }

    return value.toFixed(2).replace('.', ',');
}

function profileNumeric(values) {
    const sorted = [...values].sort((a, b) => a - b);
    const q1 = quantile(sorted, 0.25);
    const median = quantile(sorted, 0.5);
    const q3 = quantile(sorted, 0.75);
    const iqr = q3 - q1;

    // Batas Tukey 1,5 × IQR — ambang yang sama dipakai halaman Cleaning dan
    // algoritma anomali, supaya "outlier" berarti hal yang sama di seluruh aplikasi.
    const lowerFence = q1 - 1.5 * iqr;
    const upperFence = q3 + 1.5 * iqr;
    const outliers = sorted.filter(
        (value) => value < lowerFence || value > upperFence,
    );

    const min = sorted[0] ?? 0;
    const max = sorted[sorted.length - 1] ?? 0;
    const binCount = 8;
    const width = (max - min) / binCount || 1;
    const counts = new Array(binCount).fill(0);

    for (const value of sorted) {
        const index = Math.min(binCount - 1, Math.floor((value - min) / width));

        counts[index] += 1;
    }

    return {
        mean: mean(sorted),
        std: stdDev(sorted),
        min,
        max,
        q1,
        median,
        q3,
        lowerFence: Math.max(min, lowerFence),
        upperFence: Math.min(max, upperFence),
        outlierCount: outliers.length,
        outlierRatio: sorted.length ? outliers.length / sorted.length : 0,
        histogram: {
            labels: counts.map((_, index) => {
                const from = min + index * width;

                return `${formatNumber(from)}–${formatNumber(from + width)}`;
            }),
            counts,
        },
    };
}

function profileCategorical(values) {
    const tally = new Map();

    for (const value of values) {
        const key = String(value);

        tally.set(key, (tally.get(key) ?? 0) + 1);
    }

    const top = [...tally.entries()]
        .map(([value, count]) => ({
            value,
            count,
            share: values.length ? (count / values.length) * 100 : 0,
        }))
        .sort((a, b) => b.count - a.count);

    return { top };
}

function profileColumn(column, rows) {
    const raw = rows.map((row) => row[column.name]);
    const present = raw.filter((value) => value !== null && value !== undefined);
    const missingCount = raw.length - present.length;
    const unique = new Set(present.map((value) => String(value))).size;

    const base = {
        name: column.name,
        type: column.type,
        missingCount,
        missing: raw.length ? (missingCount / raw.length) * 100 : 0,
        unique,
        // Kolom teks yang seluruh nilainya unik adalah identitas baris, bukan
        // variabel yang bisa dianalisis.
        isIdentifier:
            column.type === 'text' && unique === present.length && unique > 1,
    };

    if (isNumericType(column.type)) {
        return { ...base, ...profileNumeric(present.map(Number)) };
    }

    if (column.type === 'category' || column.type === 'text') {
        return { ...base, ...profileCategorical(present) };
    }

    return base;
}

function countDuplicates(rows, columns) {
    // Kolom identitas diabaikan: dua baris dengan isi sama tetap duplikat
    // walaupun nomor transaksinya berbeda.
    const keys = columns
        .filter((column) => !column.isIdentifier)
        .map((column) => column.name);
    const seen = new Set();
    let duplicates = 0;

    for (const row of rows) {
        const signature = keys.map((key) => row[key]).join('');

        if (seen.has(signature)) {
            duplicates += 1;
        } else {
            seen.add(signature);
        }
    }

    return duplicates;
}

function buildCorrelation(rows, numericColumns) {
    const names = numericColumns.slice(0, 6).map((column) => column.name);
    const series = names.map((name) =>
        rows.map((row) => (row[name] === null ? NaN : Number(row[name]))),
    );

    return {
        columns: names,
        matrix: series.map((left) =>
            series.map((right) => Number(pearson(left, right).toFixed(2))),
        ),
    };
}

/**
 * Hasil profiling satu tabel.
 *
 * @returns {{ rowCount, columns, numeric, categorical, datetime, identifiers,
 *   correlation, duplicateRows, outlierRatio }}
 */
export function profileTable(table) {
    const columns = table.columns.map((column) =>
        profileColumn(column, table.rows),
    );

    const numeric = columns.filter(
        (column) => isNumericType(column.type) && !column.isIdentifier,
    );
    const categorical = columns.filter(
        (column) => column.type === 'category' && !column.isIdentifier,
    );
    const datetime = columns.filter((column) => column.type === 'datetime');
    const identifiers = columns.filter((column) => column.isIdentifier);

    const outlierTotal = numeric.reduce(
        (sum, column) => sum + column.outlierCount,
        0,
    );

    return {
        rowCount: table.rows.length,
        columns,
        numeric,
        categorical,
        datetime,
        identifiers,
        correlation: numeric.length >= 2 ? buildCorrelation(table.rows, numeric) : null,
        duplicateRows: countDuplicates(table.rows, columns),
        outlierRatio: numeric.length
            ? outlierTotal / (numeric.length * table.rows.length)
            : 0,
    };
}

/** Pasangan kolom numerik dengan korelasi absolut tertinggi. */
export function strongestPair(profile) {
    if (!profile.correlation) {
        return null;
    }

    const { columns, matrix } = profile.correlation;
    let best = null;

    for (let row = 0; row < columns.length; row += 1) {
        for (let col = row + 1; col < columns.length; col += 1) {
            const value = matrix[row][col];

            if (!best || Math.abs(value) > Math.abs(best.value)) {
                best = { x: columns[row], y: columns[col], value };
            }
        }
    }

    return best;
}

File: python/core/profiler.py
"""Profiling dataset: karakteristik setiap kolom.

Hasil modul ini adalah fondasi seluruh fitur lain. Rekomendasi algoritma,
pemilihan target di Machine Learning, penilaian kualitas data, dan pemilihan
grafik otomatis semuanya membaca profil yang sama, sehingga angka yang muncul di
satu halaman tidak pernah berbeda dengan halaman lain.
"""

from __future__ import annotations

import pandas as pd

from core import loader

# Batas Tukey. Ambang yang sama dipakai halaman Cleaning dan deteksi anomali
# supaya kata "outlier" berarti hal yang persis sama di seluruh aplikasi.
IQR_MULTIPLIER = 1.5

HISTOGRAM_BINS = 10
TOP_VALUES = 12


def profile_column(series: pd.Series, position: int) -> dict:
    kind = loader.column_kind(series)
    total = len(series)
    missing = int(series.isna().sum())
    non_null = series.dropna()

    profile = {
        "position": position,
        "name": str(series.name),
        "type": kind,
        "missing_count": missing,
        "missing_percent": round(missing / total * 100, 4) if total else 0.0,
        "unique_count": int(non_null.nunique()),
        "is_identifier": loader.is_identifier(series),
        "mean": None,
        "std": None,
        "min": None,
        "q1": None,
        "median": None,
        "q3": None,
        "max": None,
        "skewness": None,
        "kurtosis": None,
        "outlier_count": 0,
        "top_values": None,
        "histogram": None,
    }

    if kind in ("integer", "float") and not non_null.empty:
        profile.update(_numeric_profile(non_null))
    elif kind in ("category", "boolean", "text") and not non_null.empty:
        profile["top_values"] = _top_values(non_null)
    elif kind == "datetime" and not non_null.empty:
        profile["min"] = None
        profile["max"] = None
        profile["range"] = {
            "start": non_null.min().isoformat(),
            "end": non_null.max().isoformat(),
        }

    return profile


def _numeric_profile(values: pd.Series) -> dict:
    q1 = float(values.quantile(0.25))
    q3 = float(values.quantile(0.75))
    iqr = q3 - q1
    lower = q1 - IQR_MULTIPLIER * iqr
    upper = q3 + IQR_MULTIPLIER * iqr
    outliers = values[(values < lower) | (values > upper)]

    counts, edges = _histogram(values)

    return {
        "mean": float(values.mean()),
        "std": float(values.std()) if len(values) > 1 else 0.0,
        "min": float(values.min()),
        "q1": q1,
        "median": float(values.median()),
        "q3": q3,
        "max": float(values.max()),
        # Skewness dan kurtosis butuh minimal 3-4 titik; di bawah itu pandas
        # mengembalikan NaN yang akan menjadi null saat dibersihkan.
        "skewness": float(values.skew()) if len(values) > 2 else None,
        "kurtosis": float(values.kurtosis()) if len(values) > 3 else None,
        "outlier_count": int(len(outliers)),
        "lower_fence": float(max(values.min(), lower)),
        "upper_fence": float(min(values.max(), upper)),
        "histogram": {"counts": counts, "edges": edges},
    }


def _histogram(values: pd.Series) -> tuple:
    counts, edges = pd.cut(values, bins=HISTOGRAM_BINS, retbins=True)
    frequency = counts.value_counts(sort=False).tolist()

    return [int(item) for item in frequency], [float(edge) for edge in edges]


def _top_values(values: pd.Series) -> list:
    total = len(values)
    counts = values.astype(str).value_counts().head(TOP_VALUES)

    return [
        {
            "value": str(index),
            "count": int(count),
            "share": round(count / total * 100, 4),
        }
        for index, count in counts.items()
    ]


def correlation_matrix(frame: pd.DataFrame, columns: list, method: str = "pearson") -> dict | None:
    """Matriks korelasi antar kolom numerik."""
    if len(columns) < 2:
        return None

    matrix = frame[columns].corr(method=method, numeric_only=True)

    return {
        "columns": [str(name) for name in matrix.columns],
        "matrix": [[round(float(value), 4) for value in row] for row in matrix.values],
        "method": method,
    }


def run(frame: pd.DataFrame) -> dict:
    """Profil lengkap satu dataset."""
    columns = [
        profile_column(frame[name], position)
        for position, name in enumerate(frame.columns)
    ]

    numeric = loader.numeric_columns(frame)
    categorical = loader.categorical_columns(frame)
    datetimes = loader.datetime_columns(frame)

    # Duplikat dihitung tanpa kolom identitas: dua baris dengan isi sama tetap
    # duplikat walaupun nomor transaksinya berbeda.
    comparable = [name for name in frame.columns if not loader.is_identifier(frame[name])]
    duplicate_rows = int(frame.duplicated(subset=comparable or None).sum())

    cells = frame.shape[0] * frame.shape[1]
    missing_cells = int(frame.isna().sum().sum())
    outlier_total = sum(column["outlier_count"] for column in columns)

    return {
        "row_count": int(frame.shape[0]),
        "column_count": int(frame.shape[1]),
        "columns": columns,
        "numeric_columns": numeric,
        "categorical_columns": categorical,
        "datetime_columns": datetimes,
        "identifier_columns": [
            name for name in frame.columns if loader.is_identifier(frame[name])
        ],
        "duplicate_rows": duplicate_rows,
        "missing_cells": missing_cells,
        "missing_percent": round(missing_cells / cells * 100, 4) if cells else 0.0,
        "outlier_ratio": round(
            outlier_total / (len(numeric) * frame.shape[0]), 6
        ) if numeric and frame.shape[0] else 0.0,
        "correlation": correlation_matrix(frame, numeric[:12]),
        "preview": _preview(frame),
        "problematic_preview": _problematic_preview(frame),
    }


def _preview(frame: pd.DataFrame, rows: int = 10) -> dict:
    if frame.empty:
        return {
            "columns": [],
            "types": [],
            "rows": [],
        }

    head = frame.head(rows)

    return {
        "columns": ["Baris", *[str(name) for name in head.columns]],
        "types": ["number", *[loader.column_kind(frame[name]) for name in head.columns]],
        "rows": [
            [
                int(index) + 1,
                *[None if pd.isna(value) else str(value) for value in row],
            ]
            for index, row in zip(head.index, head.itertuples(index=False))
        ],
    }


def _problematic_preview(frame: pd.DataFrame, rows: int = 10) -> dict:
    mask = pd.Series(False, index=frame.index)

    if not frame.empty:
        mask |= frame.isna().any(axis=1)

        comparable = [name for name in frame.columns if not loader.is_identifier(frame[name])]
        if comparable:
            mask |= frame.duplicated(subset=comparable or None, keep=False)

        numeric = loader.numeric_columns(frame)
        if numeric:
            outlier_mask = pd.Series(False, index=frame.index)

            for name in numeric:
                values = frame[name]
                q1, q3 = values.quantile(0.25), values.quantile(0.75)
                iqr = q3 - q1
                lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
                outlier_mask |= (values < lower) | (values > upper)

            mask |= outlier_mask

    problematic = frame[mask]

    if problematic.empty:
        return _preview(problematic, rows)

    return _preview(problematic, rows)
